vault/client.go

```diff
@@ -1,84 +1,167 @@
+	req.SetContext(ctx)
 
-	if vaultIAMServerID != "" {
-		req.HTTPRequest.Header.Add("X-Vault-AWS-IAM-Server-ID", vaultIAMServerID)
+	if c.authConfig.IAMServerID != "" {
+		req.HTTPRequest.Header.Add("X-Vault-AWS-IAM-Server-ID", c.authConfig.IAMServerID)
 	}
 
 	if signErr := req.Sign(); signErr != nil {
-		return nil, nil, signErr
+		return signErr
 	}
 
 	headers, err := json.Marshal(req.HTTPRequest.Header)
 	if err != nil {
-		return nil, nil, err
+		return err
 	}
 
 	body, err := ioutil.ReadAll(req.HTTPRequest.Body)
 	if err != nil {
-		return nil, nil, err
+		return err
 	}
 
 	d := make(map[string]interface{})
 	d["iam_http_request_method"] = req.HTTPRequest.Method
 	d["iam_request_url"] = base64.StdEncoding.EncodeToString([]byte(req.HTTPRequest.URL.String()))
 	d["iam_request_headers"] = base64.StdEncoding.EncodeToString(headers)
 	d["iam_request_body"] = base64.StdEncoding.EncodeToString(body)
-	d["role"] = vaultAuthRole
+	d["role"] = c.authConfig.Role
 
-	logger.Println("attemping Vault login...")
-	resp, err := vaultClient.Logical().Write(fmt.Sprintf("auth/%s/login", vaultAuthProvider), d)
+	secret, err := c.VaultClient.Logical().Write(fmt.Sprintf("auth/%s/login", c.authConfig.Provider), d)
 	if err != nil {
-		return nil, nil, err
+		return err
+	}
+	if secret == nil {
+		return fmt.Errorf("got no response from the %s authentication provider", c.authConfig.Provider)
 	}
-	if resp == nil {
-		return nil, nil, fmt.Errorf("got no response from the %s authentication provider", vaultAuthProvider)
+
+	token, err := secret.TokenID()
+	if err != nil {
+		return fmt.Errorf("error reading token: %s", err)
+	}
+	c.VaultClient.SetToken(token)
+
+	err = c.updateTokenMetadata(secret)
+	if err != nil {
+		return err
 	}
 
-	token, err := parseToken(resp)
+	return nil
+}
+
+func (c *Client) renew() error {
+	secret, err := c.VaultClient.Auth().Token().RenewSelf(int(c.tokenTTL.Seconds()))
 	if err != nil {
-		return nil, nil, fmt.Errorf("error parsing token: %s", err)
+		return err
 	}
-	vaultClient.SetToken(token)
 
-	return vaultClient, config, nil
+	return c.updateTokenMetadata(secret)
 }
 
-func parseToken(resp *api.Secret) (string, error) {
+// Stores metadata about token lease that informs when to re-auth or renew.
+func (c *Client) updateTokenMetadata(secret *api.Secret) error {
 	var err error
-	token, err := resp.TokenID()
+	c.tokenTTL, err = secret.TokenTTL()
 	if err != nil {
-		return "", err
+		return err
 	}
 
-	return token, nil
+	c.tokenExpiry = time.Now().Add(c.tokenTTL)
+	c.tokenRenewable = secret.Renewable
+
+	return nil
+}
+
+// Returns true if current time is after tokenExpiry, or within 10s.
+func (c *Client) expired() bool {
+	return time.Now().Add(10 * time.Second).After(c.tokenExpiry)
+}
+
+// Returns true if tokenExpiry time is in less than 20% of tokenTTL.
+func (c *Client) shouldRenew() bool {
+	remaining := c.tokenExpiry.Sub(time.Now())
+	return c.tokenRenewable && remaining.Nanoseconds() < c.tokenTTL.Nanoseconds()/5
```

Comment:
I'm wondering if doing a token lookup self and getting the actual time left instead of storing metadata is a more reliable solution. Tracking time in the function could be subject to clock skew, especially when TTLs are small. One downside to lookups is we increase the traffic to Vault, but it's likely more reliable in the end.

Reply:
This is a good thought, and definitely worth considering. Of course the same comment also applies to the logic in `expired()` too.

On clock skew, I don't think we need to worry about cross-server clock skew because this value is calculated from a TTL duration and then only consumed within the same execution environment. Obviously that's not the same as a VM, but hopefully an execution environment's clock is more reliable at tracking a duration than the clock skew that can happen between servers.

Given that, the best we can do is eliminate the clock drift that could happen between 0 and 80% of the TTL. I don't have data, but I would hope that's normally insignificant compared to the drift due to network latency between Lambda and Vault. That network latency error could be large as we calculate the expiry time from the time we receive the response, instead of the unknown time Vault sent it. So eliminating clock drift is probably not enough gain to justify the token-lookup latency on every request, which is pure compute cost in a Lambda function.

On the topic of small TTLs, we will only ever attempt to renew if the token is not considered "expired", which means it must have at least 10s of TTL left. I think that should serve most deployments well, but thinking this through has convinced me it would be good to make that 10s grace window configurable.

Reply:
I've now added a configurable for token TTL grace period.